File: app/services/payment_service.py
import logging
from datetime import date
from typing import Any

from app.enums import PaymentMethod, PaymentType
from app.events import cash_payment_submitted, credit_purchased, payment_completed
from app.models import Credit, Membership, Payment, User
from app.repositories import CreditRepository, MembershipRepository, PaymentRepository, UserRepository
from app.services.result import ServiceResult
from app.services.settings_service import SettingsService
from app.services.sumup_service import SumUpService

logger = logging.getLogger(__name__)


class PaymentService:
    def __init__(self, processor: Any = None) -> None:
        self.processor = processor or SumUpService()

    def create_checkout(self, amount_cents: int, description: str) -> dict[str, Any] | None:
        return self.processor.create_checkout(
            amount=amount_cents,
            currency="EUR",
            description=description,
        )

    def initiate_membership_payment(self, user: User) -> ServiceResult[dict]:
        """Create payment record and SumUp checkout for membership renewal.

        On success, ``result.data`` contains::

            {"checkout_id": str, "payment_id": int, "amount": float, "description": str}
        """
        amount_cents: int = SettingsService.get("annual_membership_cost")
        description = f"Annual Membership - {user.name}"

        payment = Payment(
            user_id=user.id,
            amount_cents=amount_cents,
            currency="EUR",
            payment_type=PaymentType.MEMBERSHIP,
            payment_method=PaymentMethod.ONLINE,
            description=description,
            status="pending",
        )
        PaymentRepository.add(payment)
        PaymentRepository.save()

        checkout = self.create_checkout(amount_cents=amount_cents, description=description)

        if checkout:
            return ServiceResult.ok(
                data={
                    "checkout_id": checkout.get("id"),
                    "payment_id": payment.id,
                    "user_id": user.id,
                    "amount": amount_cents / 100.0,
                    "description": description,
                }
            )
        else:
            PaymentRepository.delete(payment)
            PaymentRepository.save()
            return ServiceResult.fail("Error creating payment. Please try again.")

    def initiate_credit_purchase(self, user: User, quantity: int) -> ServiceResult[dict]:
        """Create payment record and SumUp checkout for credit purchase.

        On success, ``result.data`` contains::

            {"checkout_id": str, "payment_id": int, "user_id": int, "quantity": int, "amount": float, "description": str}
        """
        amount_cents: int = quantity * SettingsService.get("additional_shoot_cost")
        description = f"{quantity} credits - {user.name}"

        payment = Payment(
            user_id=user.id,
            amount_cents=amount_cents,
            currency="EUR",
            payment_type=PaymentType.CREDITS,
            payment_method=PaymentMethod.ONLINE,
            description=f"{quantity} shooting credits",
            status="pending",
        )
        PaymentRepository.add(payment)
        PaymentRepository.save()

        checkout = self.create_checkout(amount_cents=amount_cents, description=description)

        if checkout:
            return ServiceResult.ok(
                data={
                    "checkout_id": checkout.get("id"),
                    "payment_id": payment.id,
                    "user_id": user.id,
                    "quantity": quantity,
                    "amount": amount_cents / 100.0,
                    "description": description,
                }
            )
        else:
            PaymentRepository.delete(payment)
            PaymentRepository.save()
            return ServiceResult.fail("Error creating payment. Please try again.")

    @staticmethod
    def initiate_cash_membership_payment(user: User) -> ServiceResult[dict]:
        """Create pending cash payment record for membership.

        On success, ``result.data`` contains::

            {"payment_id": int, "amount": float, "instructions": str}
        """
        amount_cents: int = SettingsService.get("annual_membership_cost")

        payment = Payment(
            user_id=user.id,
            amount_cents=amount_cents,
            currency="EUR",
            payment_type=PaymentType.MEMBERSHIP,
            payment_method=PaymentMethod.CASH,
            description=f"Annual Membership (Cash) - {user.name}",
            status="pending",
        )
        PaymentRepository.add(payment)
        PaymentRepository.save()

        # Emit event — handler sends confirmation email
        try:
            cash_payment_submitted.send(user_id=user.id, payment_id=payment.id)
        except Exception as e:
            logger.error(f"Failed to emit cash_payment_submitted event: {e}")

        return ServiceResult.ok(
            data={
                "payment_id": payment.id,
                "amount": amount_cents / 100.0,
                "instructions": SettingsService.get("cash_payment_instructions"),
            }
        )

    @staticmethod
    def initiate_cash_credit_purchase(user: User, quantity: int) -> ServiceResult[dict]:
        """Create pending cash payment record for credit purchase.

        On success, ``result.data`` contains::

            {"payment_id": int, "quantity": int, "amount": float, "instructions": str}
        """
        amount_cents: int = quantity * SettingsService.get("additional_shoot_cost")

        payment = Payment(
            user_id=user.id,
            amount_cents=amount_cents,
            currency="EUR",
            payment_type=PaymentType.CREDITS,
            payment_method=PaymentMethod.CASH,
            description=f"{quantity} shooting credits (Cash)",
            status="pending",
        )
        PaymentRepository.add(payment)
        PaymentRepository.save()

        # Emit event — handler sends confirmation email
        try:
            cash_payment_submitted.send(user_id=user.id, payment_id=payment.id)
        except Exception as e:
            logger.error(f"Failed to emit cash_payment_submitted event: {e}")

        return ServiceResult.ok(
            data={
                "payment_id": payment.id,
                "quantity": quantity,
                "amount": amount_cents / 100.0,
                "instructions": SettingsService.get("cash_payment_instructions"),
            }
        )

    @staticmethod
    def get_pending_cash_payment_rows() -> list[dict[str, Any]]:
        payments = PaymentRepository.get_pending_cash()
        return [{"payment": payment, "user": UserRepository.get_by_id(payment.user_id)} for payment in payments]

    @staticmethod
    def get_completed_membership_payment(user_id: int) -> Payment | None:
        return PaymentRepository.get_completed_for_user(user_id, PaymentType.MEMBERSHIP)

    @staticmethod
    def approve_cash_payment(payment_id: int) -> ServiceResult[dict[str, Any]]:
        payment = PaymentRepository.get_by_id(payment_id)
        if not payment:
            return ServiceResult.fail("Payment not found.")
        if payment.status != "pending" or payment.payment_method != PaymentMethod.CASH:
            return ServiceResult.fail("This payment cannot be approved.")

        member = UserRepository.get_by_id(payment.user_id)
        if not member:
            return ServiceResult.fail("User not found.")

        try:
            payment.mark_completed(processor="cash")
            if payment.payment_type == PaymentType.MEMBERSHIP:
                if member.membership:
                    if member.membership.status != "active":
                        member.membership.activate()
                    else:
                        expiry_date = SettingsService.calculate_membership_expiry(date.today()).date()
                        member.membership.renew(expiry_date=expiry_date)
                else:
                    expiry_date = SettingsService.calculate_membership_expiry(date.today()).date()
                    membership = Membership(
                        user_id=member.id,
                        start_date=date.today(),
                        expiry_date=expiry_date,
                        initial_credits=SettingsService.get("membership_shoots_included"),
                        purchased_credits=0,
                        status="active",
                    )
                    MembershipRepository.add(membership)
            elif payment.payment_type == PaymentType.CREDITS:
                quantity = PaymentService._credit_quantity_from_description(payment.description)
                if member.membership:
                    member.membership.add_credits(quantity)
                CreditRepository.add(Credit(user_id=member.id, amount=quantity, payment_id=payment.id))

            PaymentRepository.save()
            PaymentService._emit_payment_completed_events(payment, member)
            return ServiceResult.ok(data={"member_name": member.name}, message=f"Payment approved for {member.name}!")
        except Exception as exc:
            logger.error("Error approving payment: %s", exc)
            return ServiceResult.fail(f"Error approving payment: {exc}")

    @staticmethod
    def reject_cash_payment(payment_id: int) -> ServiceResult[dict[str, Any]]:
        payment = PaymentRepository.get_by_id(payment_id)
        if not payment:
            return ServiceResult.fail("Payment not found.")
        if payment.status != "pending" or payment.payment_method != PaymentMethod.CASH:
            return ServiceResult.fail("This payment cannot be rejected.")
        member = UserRepository.get_by_id(payment.user_id)
        payment.status = "cancelled"
        PaymentRepository.save()
        member_name = member.name if member else "user"
        return ServiceResult.ok(data={"member_name": member_name}, message=f"Payment rejected for {member_name}.")

    @staticmethod
    def _credit_quantity_from_description(description: str | None) -> int:
        if description and "shooting credits" in description.lower():
            try:
                return int(description.split()[0])
            except ValueError, IndexError:
                return 1
        return 1

    @staticmethod
    def _emit_payment_completed_events(payment: Payment, member: User) -> None:
        try:
            if payment.payment_type == PaymentType.CREDITS:
                quantity = PaymentService._credit_quantity_from_description(payment.description)
                credit_purchased.send(user_id=member.id, payment_id=payment.id, quantity=quantity)
            else:
                payment_completed.send(user_id=member.id, payment_id=payment.id, payment_type=payment.payment_type)
        except Exception:
            pass

File: app/Services/PaymentService.php
<?php

namespace App\Services;

use App\Enums\PaymentMethod;
use App\Enums\PaymentStatus;
use App\Enums\PaymentType;
use App\Mail\MembershipActivatedMail;
use App\Mail\PaymentReceiptMail;
use App\Models\Payment;
use App\Models\User;
use App\Settings\ApplicationSettings;
use Illuminate\Support\Facades\Log;
use Illuminate\Support\Facades\Mail;

class PaymentService
{
    public function __construct(
        private SumUpService $sumUpService,
        private ApplicationSettings $settings,
    ) {
    }

    /**
     * Create a checkout for payment.
     *
     * @return array<string, mixed>|null
     */
    public function createCheckout(int $amountCents, string $description): ?array
    {
        return $this->sumUpService->createCheckout($amountCents, 'EUR', $description);
    }

    /**
     * Process a payment with card details.
     *
     * @return array<string, mixed>
     */
    public function processPayment(
        string $checkoutId,
        string $cardNumber,
        string $cardName,
        string $expiryMonth,
        string $expiryYear,
        string $cvv,
    ): array {
        return $this->sumUpService->processCheckoutPayment(
            $checkoutId,
            $cardNumber,
            $cardName,
            $expiryMonth,
            $expiryYear,
            $cvv,
        );
    }

    /**
     * Initiate membership payment via online payment.
     *
     * @return array<string, mixed>
     */
    public function initiateMembershipPayment(User $user): array
    {
        $amountCents = $this->settings->annual_membership_cost;
        $description = "Annual Membership - {$user->name}";

        $payment = Payment::create([
            'user_id' => $user->id,
            'amount_cents' => $amountCents,
            'currency' => 'EUR',
            'payment_type' => PaymentType::Membership,
            'payment_method' => PaymentMethod::Online,
            'description' => $description,
            'status' => PaymentStatus::Pending,
        ]);

        $checkout = $this->createCheckout($amountCents, $description);

        if ($checkout) {
            return [
                'success' => true,
                'checkout_id' => $checkout['id'],
                'payment_id' => $payment->id,
                'amount' => $amountCents / 100.0,
                'description' => $description,
            ];
        }

        $payment->delete();
        Log::error('Failed to create checkout for membership payment', ['user_id' => $user->id]);

        return [
            'success' => false,
            'error' => 'Error creating payment. Please try again.',
        ];
    }

    /**
     * Initiate credit purchase via online payment.
     *
     * @return array<string, mixed>
     */
    public function initiateCreditPurchase(User $user, int $quantity): array
    {
        $amountCents = $quantity * $this->settings->additional_shoot_cost;
        $description = "{$quantity} shooting credits";

        $payment = Payment::create([
            'user_id' => $user->id,
            'amount_cents' => $amountCents,
            'currency' => 'EUR',
            'payment_type' => PaymentType::Credits,
            'payment_method' => PaymentMethod::Online,
            'description' => $description,
            'status' => PaymentStatus::Pending,
        ]);

        $checkout = $this->createCheckout($amountCents, "{$quantity} credits - {$user->name}");

        if ($checkout) {
            return [
                'success' => true,
                'checkout_id' => $checkout['id'],
                'payment_id' => $payment->id,
                'quantity' => $quantity,
                'amount' => $amountCents / 100.0,
                'description' => $description,
            ];
        }

        $payment->delete();
        Log::error('Failed to create checkout for credit purchase', [
            'user_id' => $user->id,
            'quantity' => $quantity,
        ]);

        return [
            'success' => false,
            'error' => 'Error creating payment. Please try again.',
        ];
    }

    /**
     * Create a pending cash payment for membership.
     */
    public function createCashMembershipPayment(User $user): Payment
    {
        $amountCents = $this->settings->annual_membership_cost;

        return Payment::create([
            'user_id' => $user->id,
            'amount_cents' => $amountCents,
            'currency' => 'EUR',
            'payment_type' => PaymentType::Membership,
            'payment_method' => PaymentMethod::Cash,
            'description' => "Annual Membership (Cash) - {$user->name}",
            'status' => PaymentStatus::Pending,
        ]);
    }

    /**
     * Create a pending cash payment for credits.
     */
    public function createCashCreditPayment(User $user, int $quantity): Payment
    {
        $amountCents = $quantity * $this->settings->additional_shoot_cost;

        return Payment::create([
            'user_id' => $user->id,
            'amount_cents' => $amountCents,
            'currency' => 'EUR',
            'payment_type' => PaymentType::Credits,
            'payment_method' => PaymentMethod::Cash,
            'description' => "{$quantity} shooting credits (Cash)",
            'status' => PaymentStatus::Pending,
        ]);
    }

    /**
     * Complete a payment and apply membership/credits.
     */
    public function completePayment(Payment $payment, ?string $transactionId = null): void
    {
        $payment->markCompleted($transactionId, 'sumup');

        $user = $payment->user;

        if ($payment->payment_type === PaymentType::Membership) {
            $this->activateMembership($user);
            $user->refresh();
            if ($user->membership) {
                Mail::to($user)->send(new MembershipActivatedMail($user, $user->membership));
            }
        } elseif ($payment->payment_type === PaymentType::Credits) {
            $this->addCredits($user, $payment);
        }

        Mail::to($user)->send(new PaymentReceiptMail($user, $payment));
    }

    /**
     * Activate or renew membership for a user.
     */
    private function activateMembership(User $user): void
    {
        $membership = $user->membership;

        if ($membership) {
            $membership->renew($this->settings->membership_shoots_included);
        } else {
            $user->membership()->create([
                'start_date' => now(),
                'expiry_date' => $this->calculateExpiryDate(),
                'initial_credits' => $this->settings->membership_shoots_included,
                'purchased_credits' => 0,
                'status' => \App\Enums\MembershipStatus::Active,
            ]);
        }
    }

    /**
     * Add credits to user based on payment.
     */
    private function addCredits(User $user, Payment $payment): void
    {
        // Calculate quantity from payment amount
        $quantity = (int) ($payment->amount_cents / $this->settings->additional_shoot_cost);

        $membership = $user->membership;
        if ($membership) {
            $membership->addCredits($quantity);
        }

        // Create credit record
        $user->credits()->create([
            'amount' => $quantity,
            'payment_id' => $payment->id,
        ]);
    }

    /**
     * Calculate membership expiry date based on settings.
     */
    private function calculateExpiryDate(): \Carbon\Carbon
    {
        $now = now();
        $yearStartMonth = $this->settings->membership_year_start_month;
        $yearStartDay = $this->settings->membership_year_start_day;

        // Create this year's start date
        $thisYearStart = $now->copy()->setMonth($yearStartMonth)->setDay($yearStartDay)->startOfDay();

        // If we're past this year's start, expiry is next year's start
        if ($now->gte($thisYearStart)) {
            return $thisYearStart->addYear();
        }

        // Otherwise, expiry is this year's start
        return $thisYearStart;
    }
}
